src/main/java/net/neoforged/neoforge/common/crafting/NBTIngredient.java

```diff
@@ -68,33 +91,64 @@ public boolean synchronizeWithContents() {
     }
 
     private static boolean compareStacksWithNBT(ItemStack left, ItemStack right) {
-        return left.getItem() == right.getItem() && NbtUtils.compareNbt(left.getTag(), right.getTag(), true);
+        return left.is(right.getItem()) && NbtUtils.compareNbt(left.getTag(), right.getTag(), true) && NbtUtils.compareNbt(left.serializeAttachments(), right.serializeAttachments(), true);
     }
 
     /**
      * Creates a new ingredient matching any item from the list, containing the given NBT
      */
     public static NBTIngredient of(boolean strict, CompoundTag nbt, ItemLike... items) {
-        return new NBTIngredient(Arrays.stream(items).map(ItemLike::asItem).collect(Collectors.toSet()), nbt, strict);
+        Objects.requireNonNull(nbt, "NBT Ingredient requires NBT");
+        return new NBTIngredient(Arrays.stream(items).map(ItemLike::asItem).collect(Collectors.toSet()), nbt, null, strict);
+    }
+
+    /**
+     * Creates a new ingredient matching any item from the list, containing the given NBT
+     */
+    public static NBTIngredient of(boolean strict, CompoundTag nbt, CompoundTag attachmentNbt, ItemLike... items) {
+        Objects.requireNonNull(nbt, "NBT Ingredient requires NBT");
+        Objects.requireNonNull(attachmentNbt, "NBT Ingredient requires attachment NBT");
+        return new NBTIngredient(Arrays.stream(items).map(ItemLike::asItem).collect(Collectors.toSet()), nbt, attachmentNbt, strict);
+    }
+
+    /**
+     * Creates a new ingredient matching any item from the list, containing the given NBT
+     */
+    public static NBTIngredient ofAttachment(boolean strict, CompoundTag attachmentNbt, ItemLike... items) {
+        Objects.requireNonNull(attachmentNbt, "NBT Ingredient requires attachment NBT");
+        return new NBTIngredient(Arrays.stream(items).map(ItemLike::asItem).collect(Collectors.toSet()), null, attachmentNbt, strict);
     }
 
     /**
      * Creates a new ingredient matching the given item, containing the given NBT
      */
     public static NBTIngredient of(boolean strict, ItemStack stack) {
-        return new NBTIngredient(Set.of(stack.getItem()), stack.getOrCreateTag(), strict);
+        CompoundTag attachmentNbt = stack.serializeAttachments();
+        //Only force create a tag if there is no serializable attachments
+        CompoundTag nbt = attachmentNbt == null ? stack.getOrCreateTag() : stack.getTag();
```

Comment:
This overload is always safe to use regardless of the stack - don't we want to let this go through the check as well?